### ISSUE TITLE
New rule suggestion: no-global-event

### ISSUE THREAD

**Issue by mzgol** [eslint/eslint#4710]
Relying on `event` being global in event handlers may cause the page to not work in some browsers as it's a non-standard feature implemented fully mainly in Trident/EdgeHTML & WebKit/Blink. Many people would prefer to make sure one does *not* use the global variable and instead uses the parameter to the event handler as you can see in the discussion at #3964.

Since removing `event` from browser globals was rejected but many people would still prefer to not use this global how about creating the `no-global-event` rule that would make sure people don't use the global? I'd expect this rule to forbid using `window.event`, `global.event` in all cases and `event` if it's not defined in the current scope or one of its ancestors.

**Comment by eslintbot** [comment]
@mzgol Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by mzgol** [comment]
> When the rules will warn. Include a description as well as sample code.

The rule would warn in three cases:
1. When it encounters `global.event` anywhere.
1. When it encounters `window.event` anywhere.
1. When it encounters `event` used as a variable but not declared in current scope or any of its parents. For example, these patterns would result in an error:
```js
alert(window.event); // `window.event` not allowed
```
```js
function handler() {
    alert(event); // implicit global `event`
}
```

These would not result in an error:
```js
var event = 'foo';
function handler() {
    alert(event);
}
```
```js
function outer() {
    var event = 'foo';
    function middle() {
        function inner() {
            alert(event);
        }
    }
}
```

> Whether the rule prevents an error or is stylistic.

In some cases it might be purely-stylistic but in most it would prevent an error: the handler throwing in Firefox and some other browsers.

> Why the rule should be in the core instead of creating a custom rule.

1. Because it prevents an error in some browsers
1. Because if you move the code from a Node.js script to a browser one you might forget to test it in all browsers.
1. Because it's a common problem for people testing in Chrome & IE: they assume the code should work in Firefox since it works in these two browsers, they don't expect those two browsers might share a non-standard feature.

> Are you willing to create the rule yourself?

Yes, I can do it.

**Comment by nzakas** [comment]
This rule is too specific to be included in the core. We are exploring creating a rule that allows you to configure which globals you want to flag: https://github.com/eslint/eslint/issues/4488

**Comment by nzakas** [comment]
Oops, I meant this one: https://github.com/eslint/eslint/issues/3966

@mzgol if you'd like to submit a PR for that, we'd welcome it.

**Comment by axelson** [comment]
Note, you can use the `no-restricted-globals` rule now: http://eslint.org/docs/rules/no-restricted-globals
